File: telemetry/src/main/java/org/strykeforce/thirdcoast/telemetry/util/SawtoothSignalGenerator.java
package org.strykeforce.thirdcoast.telemetry.util;

class SawtoothSignalGenerator extends SignalGenerator {

  public SawtoothSignalGenerator(
      double frequency, double phase, double amplitude, double offset, double invert) {
    super(frequency, phase, amplitude, offset, invert);
  }

  @Override
  double getValue(double time) {
    double t = frequency * time + phase;
    double val = 2.0 * (t - Math.floor(t + 0.5));
    return invert * amplitude * val + offset;
  }

  @Override
  public String toString() {
    return "Sawtooth at " + super.toString();
  }
}

File: src/main/kotlin/org/strykeforce/telemetry/util/SawtoothSignalGenerator.kt
package org.strykeforce.telemetry.util

internal class SawtoothSignalGenerator(
    frequency: Double, phase: Double, amplitude: Double, offset: Double, invert: Double
) : SignalGenerator(frequency, phase, amplitude, offset, invert) {

    override fun getValue(time: Double): Double {
        val t = frequency * time + phase
        val `val` = 2.0 * (t - Math.floor(t + 0.5))
        return invert * amplitude * `val` + offset
    }

    override fun toString(): String {
        return "Sawtooth at " + super.toString()
    }
}
